### ISSUE TITLE
Mark Map's Cell Surfaces as Dirty More Accurately

### ISSUE THREAD

**Issue by Chris3606** [thesadrogue/TheSadRogue.Integration#39]
# Background
Currently, the map rendering code marks all surfaces that are rendering it as dirty each time a terrain cell has its appearance changed.  This is functional, but inefficient; it's only necessary to mark a cell surface as dirty if the cell that was changed is inside that surface's viewport.  Therefore, the current code is incorrectly causing a re-render for each map renderer, on each frame where the view did not move, but a piece of terrain outside that map renderer's viewport changed.

Ideally, when a terrain appearance is changed, surfaces would only be marked as needing a re-render if that terrain was inside their viewport.  Unfortunately, this is currently difficult to implement, because the `ColoredGlyph.IsDirtySet` event only gives you access to the colored glyph itself (via `sender`), and the `ColoredGlyph` on its own does not know its own position.

# Proposal 1
SadConsole intentionally does not have a position field on `ColoredGlyphs` to conserve memory and overhead for general use cases, where most of the time it is unnecessary.  This is understandable, and in light of this I propose the following workaround to enable this optimization of the integration library:

1. create a class like so:
```CS
// This is a colored glyph class specifically designed for appearances of terrain.  It knows the terrain object
// that it is representing the appearance for, and via that, has access to its position.
public class TerrainAppearance : ColoredGlyph
{
public readonly RogueLikeCell Terrain;

// All ColoredGlyph constructors, but with added required parameter terrain which is set to the above field
}
```

2. REMOVE any and all constructors in `RogueLikeCell` that take a `ColoredGlyph` directly.

3. Modify remaining constructors to construct a `TerrainAppearance` object instead of `ColoredGlyph`, passing `this` in as the `TerrainAppearance.parent` parameter

4. Modify the `IsDirtySet` event handler to simply cast its `sender` to `TerrainApperance`, at which point it knows the position.

Note that this does not change the public interface in any breaking way; either the `Appearance` field can be left as type `ColoredGlyph`, or changed to `TerrainApperance`; in either case, both are `ColoredGlyph` subclasses so no functionality is lost from the user perspective.

# Proposal 2
A simpler way to implement this is to simply bring back the `AppearanceChanged` event on `RogueLikeCell`, and automatically fire that event each time `IsDirtySet` is fired.  It may slow things down because it's adding another event into the chain that happens when appearances are changed, completely solves the issue, as `AppearanceChanged` is an event on the `RogueLikeCell` object (which knows its position).

**Comment by Chris3606** [comment]
I think I currently prefer proposal 2.  The slow-down should be limited because this only ever happens when terrain is changed, and further it only happens once per frame per cell, since `IsDirtySet` will only fire at most once per frame per cell.

I am concerned about the overhead of the extra call if someone implements something involving terrain animation, however.  Homicide Detective's wind animation makes a good example of this, and might make a good test candidate, developer willing.

@fcheadle @Thraka ?

**Comment by Chris3606** [comment]
Note: Edited above comments to more precisely reflect my current stance.

**Comment by fcheadle** [comment]
Since `RogueLikeCell` is a `GameObject`, I don't think it makes much sense to require it for the proposed `TerrainAppearance` as opposed to giving `TerrainAppearance` a plain-old `GameObject`.

What might more more sense in this case is re-writing `RogueLikeCell` so that it inherits from `ColoredGlyph` (and not `GameObject`). Just thinking out loud here.

**Comment by Chris3606** [comment]
> I don't think it makes much sense to require it for the proposed TerrainAppearance

@fcheadle If I may ask, why not?  The only allowed terrain type in any map provided by the integration library is `RogueLikeCell`, so this isn't asserting anything that's not already run-time enforced, in my opinion.  Granted, you could implement the `Terrain` field of `TerrainAppearance` as `GameObject`, but as far as I can tell this is undesirable, as again no object that is not of type `RogueLikeCell` is allowed to reside on the terrain layer of any map provided by the library anyway.

> What might more more sense in this case is re-writing RogueLikeCell so that it inherits from ColoredGlyph (and not GameObject).

This is definitely a third option.  I have two concerns with this one:
1. The code to implement `GameObject` is comparatively non-trivial.  As such, I'm not the hugest fan of copy-pasting the entire implementation simply to gain the inheritance structure desired for this one optimization.
2. `RogueLikeCell` would have no `Appearance` property, which would place it at odds with `RogueLikeEntity` (as `Entity` has an `Appearance` property).  The  interface consistency gained by having the property may be worth considering.

**Comment by fcheadle** [comment]
Walking through the proposed solution...

`RogueLikeCell : GameObject` contains a `ColoredGlyph` named `Appearance`. 
Proposal number one would create a new class, `RogueLikeTerrain : ColoredGlyph`, which contains a `RogueLikeCell` named `Cell`.

Which class should we send to `SetTerrain()`? `SetTerrain(rogueLikeTerrain);` `SetTerrain(rogueLikeTerrain.Cell);`?

**Comment by Chris3606** [comment]
I think the names used have caused some confusion.  Proposal 1 refers to nothing more than creating a subclass of `ColoredGlyph` and effectively changing `RogueLikeCell.Appearance` to a value of the subclass type. You would still pass exactly the same thing to `SetTerrain`, no other types would be modified, other than to accomodate the subclass at construction.

**Comment by Chris3606** [comment]
A potential proposal 1 implementation is drafted up as https://github.com/thesadrogue/TheSadRogue.Integration/pull/42 in an effort to alleviate confusion.  Some elements were modified/simplified slightly from the original proposal, but the principal remains the same.

**Comment by Thraka** [comment]
Why can't a read-only position be added to the map tile that is set on the constructor according to where the tile lands on the map?

**Comment by Chris3606** [comment]
The map tiles are `RogueLikeCells`, which are game objects and already know their position.  That isn't sufficient, because to mark surfaces as dirty when appearance changes, we must hook the `IsDirtySet` event; but that event is on the _appearance_, not the cell itself; so when the event fires, the only context you have is the appearance object that was changed, which isn't sufficient to get you the position.  This is what proposal 1 changes.

**Comment by Thraka** [comment]
Gotcha!

**Comment by Chris3606** [comment]
Moved to alpha1 since we have a draft solution for this and it is a relatively significant optimization.

**Comment by Chris3606** [comment]
Fixed by #42
